### ISSUE TITLE
Image Adjustments API doesn't clearly specify input range

### ISSUE THREAD

**Issue by ahundt** [tensorflow/tensorflow#15722]
### Describe the problem

There is a documentation issue with a possible corresponding tf.keras bug.
The [tf image adjustments guide](https://www.tensorflow.org/api_guides/python/image#Image_Adjustments) doesn't document the inputs. It appears they can be in a `[0, 1]` range or `[0, MAX]` based on comments in tf.slim where the relevant APIs are used. 

This may have led to preprocessing bugs in other utilities such as keras and tf.keras, which believe tf expects the range `[-1, 1]` see https://github.com/keras-team/keras/issues/8916, the following includes additional details from that issue:

It appears Keras' imagenet image preprocessing may be inconsistent with how it is done in tf, in particular keras sets values to `[-1, 1]` while in tf the expected range is `[0, 1]`.

- [slim/preprocessing/inception_preprocessing.py](https://github.com/tensorflow/models/blob/master/research/slim/preprocessing/inception_preprocessing.py#L284) 
    - notes `If dtype is tf.float32 then the range should be [0, 1]`
- [keras preprocess_input()](https://github.com/keras-team/keras/blob/master/keras/applications/imagenet_utils.py#L72) 
    - notes `tf: will scale pixels between -1 and 1, sample-wise`.

Here is the key doc from Keras' `preprocess_input()`:

```python

def preprocess_input(x, data_format=None, mode='caffe'):
    """Preprocesses a tensor encoding a batch of images.
    # Arguments
        x: input Numpy or symoblic tensor, 3D or 4D.
        data_format: data format of the image tensor.
        mode: One of "caffe", "tf".
            - caffe: will convert the images from RGB to BGR,
                then will zero-center each color channel with
                respect to the ImageNet dataset,
                without scaling.
            - tf: will scale pixels between -1 and 1,
                sample-wise.
    # Returns
        Preprocessed tensor.
    """
```

Here are the key docs from the tf slim function `preprocess_for_train()` which specify a [0, 1] range:

```python
def preprocess_for_train(image, height, width, bbox,
                         fast_mode=True,
                         scope=None,
                         add_image_summaries=True):
  """Distort one image for training a network.
  Distorting images provides a useful technique for augmenting the data
  set during training in order to make the network invariant to aspects
  of the image that do not effect the label.
  Additionally it would create image_summaries to display the different
  transformations applied to the image.
  Args:
    image: 3-D Tensor of image. If dtype is tf.float32 then the range should be
      [0, 1], otherwise it would converted to tf.float32 assuming that the range
      is [0, MAX], where MAX is largest positive representable number for
      int(8/16/32) data type (see `tf.image.convert_image_dtype` for details).
    height: integer
    width: integer
    bbox: 3-D float Tensor of bounding boxes arranged [1, num_boxes, coords]
      where each coordinate is [0, 1) and the coordinates are arranged
      as [ymin, xmin, ymax, xmax].
    fast_mode: Optional boolean, if True avoids slower transformations (i.e.
      bi-cubic resizing, random_hue or random_contrast).
    scope: Optional scope for name_scope.
    add_image_summaries: Enable image summaries.
  Returns:
    3-D float Tensor of distorted image used for training with range [-1, 1].
  """
```

side note: 

- https://github.com/tensorflow/models/issues/2217 seems relevant
- https://github.com/keras-team/keras/issues/8916 is the corresponding keras issue

### System information

Not relevant, this is a documentation + input value range issue.

### Source code / logs

relevant links + code included above

**Comment by asimshankar** [comment]
@fchollet @anj-s : Mind taking a look?

**Comment by ahundt** [comment]
I believe I've got it working, and it seems:

1. expected input range of the image augmentation functions is [0, 1]
2. subtract 0.5 from that output and multiply by 2 to get a range of [-1, 1]
3. The expected input range when you actually feed the data to the various pretrained neural networks in the model zoo is [-1, 1].

Perhaps this can be noted clearly in these APIs?

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activityand the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by ahundt** [comment]
If someone could just point me to the location in the code, I could at least write a small docstring.

**Comment by anj-s** [comment]
@ahundt I am currently working on verifying which APIs will need this doc change. Will send a PR asap. Thanks for the clear explanation above.

**Comment by tensorflowbutler** [comment]
A member of the TensorFlow organization has replied after the stat:awaiting tensorflower label was applied.

**Comment by parag2489** [comment]
@ahundt Why will you call [tf.image.per_image_standardization()](https://www.tensorflow.org/api_docs/python/tf/image/per_image_standardization) to bring the image from `[0, 255]` to `[0, 1]`? Why shouldn't we just cast the image as `float` first and then divide by `255`? Then subtract 0.5 and multiply by 2. You will get an image that has values lying in `[-1, 1]`. This is followed in Keras Inception preprocessing as seen [here](https://stackoverflow.com/questions/44341258/preprocessing-function-of-inception-v3-in-keras), [here](https://github.com/keras-team/keras/issues/5416) and [here](https://github.com/fchollet/deep-learning-models/blob/ccd0eb24996b4cbff4231b90cd44b057c0b20f14/inception_v3.py#L391).

**Comment by ahundt** [comment]
@parag2489 good question. I may have it incorrect and that's why I created this issue hoping to clarify things. It seems there are a couple styles of preprocessing:

```
      'cifarnet': cifarnet_preprocessing,
      'inception': inception_preprocessing,
      'resnet_v1_200': vgg_preprocessing,
```

As found in https://github.com/tensorflow/models/issues/2217

**Comment by ahundt** [comment]
@parag2489 Thanks for your comments. I re-read several places where preprocessing is done and it seems you're right, dividing by 255 is the most common image preprocessing to do in tf. I've updated https://github.com/tensorflow/tensorflow/issues/15722#issuecomment-354625196 accordingly. 

However, It is important to note `img/255` is not the only valid approach, as you can see from [keras' preprocess_input function](https://github.com/keras-team/keras/blob/master/keras/applications/imagenet_utils.py#L149), which has improved since my initial post. There are options for 'tf', 'torch', and 'caffe' style preprocessing which all vary. The most important conclusion is to use the same preprocessing algorithm you trained your weights on excluding the image augmentation steps.

As for the tf documentation here I'm hopeful it can be updated to capture my outline, so others can be saved the trouble. 

@anj-s just wanted to ping you, any availability on this?

**Comment by anj-s** [comment]
ahundt@ sorry for the delay, I haven't had a chance to get to this yet. However you can update the doc string [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/image_ops_impl.py#L847) and open a PR. 

Both types of image processing are consistent from what I see.  
@asimshankar  Do you know the recommended approach?

**Comment by aminemosbah** [comment]
hi 
great post,
i m trying to run balloon detection API Mask_RCNN, and i get stuck with this error 

input that isn't a symbolic tensor

any suggestion to work this out?

**Comment by sachinprasadhs** [comment]
Hi, The document link which you have mentioned is broken. 
You can have a look into the latest documentation on [tf.image](https://www.tensorflow.org/api_docs/python/tf/image).
Also, tf.keras.preprocessing will be deprecated in future Tensorflow version, for more details please check [here](https://www.tensorflow.org/api_docs/python/tf/keras/preprocessing?version=nightly). Thanks!
